packages/botonic-nlp/src/preprocess/vocabulary.ts

```diff
@@ -0,0 +1,46 @@
+import { Dataset } from '../dataset/dataset'
+import { flatten, unique } from '../utils/array-utils'
+import { PADDING_TOKEN } from './constants'
+import { Preprocessor } from './preprocessor'
+
+export class Vocabulary {
+  readonly tokens: string[]
+  readonly length: number
+
+  constructor(tokens: string[]) {
+    this.tokens = unique(tokens)
+    this.length = this.tokens.length
+  }
+
+  static fit(dataset: Dataset, preprocessor: Preprocessor): Vocabulary {
```

Comment:
pardon my nlp ignoreance, but "fit" sounds like a ML concept. is it?
A typical factory method name is fromXXX.What about fromDataset?
however, what about moving the method to Dataset.extractVocabulary?
In case of doubt in wondering where to place a function:
* Each class should only have 1 responsibility. To which one does the function semantics fit the best?
* put the function in the class which accesses data from a class instance
* Avoid low level classes to depend on high level classes (to avoid circular dependencies)